File: 백준/Gold/24229. 모두싸인 출근길/모두싸인 출근길.py
"""
1. 이을 수 있는 판자는 모두 잇기
2. 판자마다 최대한 멀리 갈 수 있는 지점 찾기
3. 어떤 판자에서 시작했을 때 그 판자에서 최대한 멀리 갈 수 있는 지점보다 시작 점이 작은 판자들에서 또 최대한 멀리 갈 수 있는 판자들 구하기

틀림
합친 판자가 하나인 경우도 생각해야함

멀리까지 갈 수 있다고 해서 그게 답은 아닐 수 있음
"""
import sys

input = lambda: sys.stdin.readline().strip()

n = int(input())
lst = []
for _ in range(n):
    lst.append(list(map(int, input().split())))

lst = sorted(lst)
# 판자 합치기
lst_com = []
visited = [False] * n
for i in range(n):
    if visited[i]:
        continue
    start = lst[i][0]
    end = lst[i][1]
    for j in range(1, n):
        if i + j >= n:
            break
        if lst[i + j][0] <= end:
            visited[i + j] = True
            if end < lst[i + j][1]:
                end = lst[i + j][1]
        else:
            break
    lst_com.append([start, end, end - start + end])

answer = lst_com[0][1]
max_dis = lst_com[0][2]
for i in range(1, len(lst_com)):
    if lst_com[i][0] <= max_dis:
        max_dis = max(max_dis, lst_com[i][2])
        answer = max(answer, lst_com[i][1])

print(answer)

File: 백준/Gold/24229. 모두싸인 출근길/모두싸인 출근길.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static StringBuilder sb = new StringBuilder();

    public static void main(String[] args) throws IOException {

        int N = Integer.parseInt(br.readLine());
        int[][] lst = new int[N][2];

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            lst[i][0] = Integer.parseInt(st.nextToken());
            lst[i][1] = Integer.parseInt(st.nextToken());
        } // 입력완료

        Arrays.sort(lst, Comparator.comparingInt((int[] o) -> o[0]).thenComparingInt(o -> o[1]));
        List<int[]> lstCom = new ArrayList<>();
        boolean[] visited = new boolean[N];

        // 배열 합치기
        for (int i = 0; i < N; i++) {
            if (visited[i])
                continue;
            int start = lst[i][0];
            int end = lst[i][1];
            for (int j = 1; j < N; j++) {
                if (i + j >= N)
                    break;
                if (lst[i + j][0] <= end) {
                    visited[i + j] = true;
                    if (end < lst[i + j][1]){
                        end = lst[i + j][1];
                    }
                } else {
                    break;
                }
            }
            lstCom.add(new int[] {start, end, end - start + end});
        } // end of for i

        int answer = lstCom.get(0)[1];
        int maxDis = lstCom.get(0)[2];
        for (int i = 1; i < lstCom.size(); i++) {
            if (lstCom.get(i)[0] <= maxDis) {
                answer = Math.max(answer, lstCom.get(i)[1]);
                maxDis = Math.max(maxDis, lstCom.get(i)[2]);
            }
        } // end of for i

        System.out.println(answer);
    } // end of main

} // end of class